From 4d481b5fab653fc150502eaabcb2845ba326f680 Mon Sep 17 00:00:00 2001
From: Sven Oliver Moll <svolli@svolli.de>
Date: Fri, 23 Jan 2026 19:44:06 +0100
Subject: [PATCH] Add compile option ENABLE_FLAT_MEMORY to optimize memory
 allocation

ENABLE_FLAT_MEMORY will call a single alloc* for a full image of pixel
data, instead of one alloc per line of pixel data. Pointers to lines will
still be set accordingly. All test cases succeeded.

autotools: --enable-flat-memory
cmake: -DENABLE_FLAT_MEMORY

Default is off to ensure maximum compatibility.

* malloc or calloc where applicable

Signed-off-by: Sven Oliver Moll <svolli@svolli.de>
---
 CMakeLists.txt        |  2 ++
 configure.ac          | 15 +++++++++++++
 docs/README.CMAKE     |  1 +
 src/config.h.cmake    | 16 ++++++++++----
 src/gd.c              | 50 ++++++++++++++++++++++++++++++++++++++++++-
 src/gd_topal.c        | 14 ++++++++++++
 windows/msys/Makefile |  2 +-
 7 files changed, 94 insertions(+), 6 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6d57d58be..0a76303df 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -10,6 +10,7 @@ SET(CMAKE_MODULE_PATH "${GD_SOURCE_DIR}/cmake/modules")
 include(gd)
 
 OPTION(ENABLE_GD_FORMATS "Enable GD image formats" 0)
+OPTION(ENABLE_FLAT_MEMORY "Enable optimized memory allocation" 0)
 OPTION(ENABLE_PNG "Enable PNG support" 0)
 OPTION(ENABLE_LIQ "Enable libimagequant support" 0)
 OPTION(ENABLE_JPEG "Enable JPEG support" 0)
@@ -204,6 +205,7 @@ else (USE_EXT_GD)
 	endif (ENABLE_RAQM)
 
 	SET(ENABLE_GD_FORMATS ${ENABLE_GD_FORMATS})
+	SET(ENABLE_FLAT_MEMORY ${ENABLE_FLAT_MEMORY})
 
 	if (FREETYPE_FOUND)
 		INCLUDE_DIRECTORIES(${FREETYPE_INCLUDE_DIRS})
diff --git a/configure.ac b/configure.ac
index c0d3ecd9c..4b31ea6c4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -116,6 +116,20 @@ fi
 AC_DEFINE_UNQUOTED([ENABLE_GD_FORMATS], [$gd_ac_value], [Whether to support gd image formats])
 AM_CONDITIONAL([ENABLE_GD_FORMATS], test "$gd_enable_gd_formats" = yes)
 
+dnl Should this be turned on in future releases?
+AC_MSG_CHECKING([whether to use flat memory model for image lines])
+AC_ARG_ENABLE([flat-memory],
+  [AS_HELP_STRING([--enable-flat-memory], [Enable flat memory implementation for image pixels])],
+  [gd_enable_flat_memory=$enableval],
+  [gd_enable_flat_memory=no])
+if test "$gd_enable_flat_memory" = yes; then
+  gd_ac_value=1
+else
+  gd_ac_value=0
+fi
+AC_DEFINE_UNQUOTED([ENABLE_FLAT_MEMORY], [$gd_ac_value], [Wheater to use one big memory chunk for pixels])
+AM_CONDITIONAL([ENABLE_FLAT_MEMORY], test "gd_enable_flat_memory" = yes)
+
 dnl Helper macro for working with external libraries.
 dnl GD_LIB_CHECK([SYM], [FEATURE], [name], [...test...])
 dnl $1 - upper case symbol
@@ -348,6 +362,7 @@ AC_MSG_RESULT([
    Support for liq library:          $gd_with_LIBIMAGEQUANT
    Support for complex text:         $gd_with_LIBRAQM
    Support for pthreads:             $ax_pthread_ok
+   Use of flat memory allocation:    $gd_enable_flat_memory
 ])
 
 AC_CONFIG_FILES([Makefile
diff --git a/docs/README.CMAKE b/docs/README.CMAKE
index c9f03827e..43b619845 100644
--- a/docs/README.CMAKE
+++ b/docs/README.CMAKE
@@ -11,6 +11,7 @@ $ make
 
 The available options are:
 ENABLE_GD_FORMATS=1
+ENABLE_FLAT_MEMORY=1
 ENABLE_PNG=1
 ENABLE_LIQ=1
 ENABLE_JPEG=1
diff --git a/src/config.h.cmake b/src/config.h.cmake
index f6f54e2cf..49dbba6ae 100644
--- a/src/config.h.cmake
+++ b/src/config.h.cmake
@@ -3,6 +3,9 @@
 /* Define is you are building for Win32 API */
 #cmakedefine BGDWIN32
 
+/* Wheater to use one big memory chunk for pixels */
+#cmakedefine01 ENABLE_FLAT_MEMORY
+
 /* Whether to support gd image formats */
 #cmakedefine01 ENABLE_GD_FORMATS
 
@@ -45,6 +48,9 @@
 /* Define if you have jpeg */
 #cmakedefine HAVE_LIBJPEG
 
+/* Define to 1 if you have the 'm' library (-lm). */
+#cmakedefine HAVE_LIBM
+
 /* Define if you have png */
 #cmakedefine HAVE_LIBPNG
 
@@ -63,9 +69,6 @@
 /* Define if you have zlib */
 #cmakedefine HAVE_LIBZ
 
-/* Define to 1 if you have the <memory.h> header file. */
-#cmakedefine HAVE_MEMORY_H
-
 /* Define if OpenMP is enabled */
 #cmakedefine HAVE_OPENMP
 
@@ -78,6 +81,9 @@
 /* Define to 1 if you have the <stdint.h> header file. */
 #cmakedefine HAVE_STDINT_H
 
+/* Define to 1 if you have the <stdio.h> header file. */
+#cmakedefine HAVE_STDIO_H
+
 /* Define to 1 if you have the <stdlib.h> header file. */
 #cmakedefine HAVE_STDLIB_H
 
@@ -131,7 +137,9 @@
    your system. */
 #cmakedefine PTHREAD_CREATE_JOINABLE
 
-/* Define to 1 if you have the ANSI C header files. */
+/* Define to 1 if all of the C89 standard headers exist (not just the ones
+   required in a freestanding environment). This macro is provided for
+   backward compatibility; new code need not use it. */
 #cmakedefine STDC_HEADERS
 
 /* Version number of package */
diff --git a/src/gd.c b/src/gd.c
index 768fd2065..4071db9bf 100644
--- a/src/gd.c
+++ b/src/gd.c
@@ -211,6 +211,17 @@
 	im->brush = 0;
 	im->tile = 0;
 	im->style = 0;
+#if ENABLE_FLAT_MEMORY
+   im->pixels[0] = (unsigned char *) gdCalloc (sx*sy, sizeof (unsigned char));
+   if (!im->pixels[0]) {
+		gdFree(im->pixels);
+		gdFree(im);
+		return NULL;
+	}
+	for (i = 1; (i < sy); i++) {
+		im->pixels[i] = im->pixels[0] + (i*sx);
+	}
+#else
 	for (i = 0; (i < sy); i++) {
 		/* Row-major ever since gd 1.3 */
 		im->pixels[i] = (unsigned char *) gdCalloc (sx, sizeof (unsigned char));
@@ -224,6 +235,7 @@
 		}
 
 	}
+#endif
 	im->sx = sx;
 	im->sy = sy;
 	im->colorsTotal = 0;
@@ -318,6 +330,17 @@
 	im->brush = 0;
 	im->tile = 0;
 	im->style = 0;
+#if ENABLE_FLAT_MEMORY
+	im->tpixels[0] = (int *) gdCalloc (sx*sy, sizeof (int));
+   if (!im->tpixels[0]) {
+		gdFree(im->tpixels);
+		gdFree(im);
+		return NULL;
+	}
+	for (i = 1; (i < sy); i++) {
+		im->tpixels[i] = im->tpixels[0] + (i*sx);
+   }
+#else
 	for (i = 0; (i < sy); i++) {
 		im->tpixels[i] = (int *) gdCalloc (sx, sizeof (int));
 		if (!im->tpixels[i]) {
@@ -332,6 +355,7 @@
 			return 0;
 		}
 	}
+#endif
 	im->sx = sx;
 	im->sy = sy;
 	im->transparent = (-1);
@@ -388,6 +412,16 @@
 
 BGD_DECLARE(void) gdImageDestroy (gdImagePtr im)
 {
+#if ENABLE_FLAT_MEMORY
+	if (im->pixels) {
+		gdFree (im->pixels[0]);
+		gdFree (im->pixels);
+	}
+	if (im->tpixels) {
+		gdFree (im->tpixels[0]);
+		gdFree (im->tpixels);
+	}
+#else
 	int i;
 
 	if (im->pixels) {
@@ -402,6 +436,7 @@
 		}
 		gdFree (im->tpixels);
 	}
+#endif
 	if (im->polyInts) {
 		gdFree (im->polyInts);
 	}
@@ -4610,17 +4645,25 @@
 		if (src->tpixels == NULL) {
 			return 0;
 		}
-
+#if ENABLE_FLAT_MEMORY
+		src->tpixels[0] = (int *) gdMalloc(sx * sy * sizeof(int));
+		if (src->tpixels[0] == NULL) {
+			goto clean_on_error;
+		}
+#endif
 		for (y = 0; y < sy; y++) {
 			const unsigned char *src_row = src->pixels[y];
 			int * dst_row;
 
 			/* no need to calloc it, we overwrite all pxl anyway */
+#if ENABLE_FLAT_MEMORY
+			src->tpixels[y] = src->tpixels[0] + (y*sx);
+#else
 			src->tpixels[y] = (int *) gdMalloc(sx * sizeof(int));
 			if (src->tpixels[y] == NULL) {
 				goto clean_on_error;
 			}
-
+#endif
 			dst_row = src->tpixels[y];
 			for (x = 0; x < sx; x++) {
 				const unsigned char c = *(src_row + x);
@@ -4634,9 +4677,13 @@
 	}
 
 	/* free old palette buffer (y is sy) */
+#if ENABLE_FLAT_MEMORY
+	gdFree(src->pixels[0]);
+#else
 	for (yy = 0; yy < y; yy++) {
 		gdFree(src->pixels[yy]);
 	}
+#endif
 	gdFree(src->pixels);
 	src->trueColor = 1;
 	src->pixels = NULL;
diff --git a/src/gd_topal.c b/src/gd_topal.c
index af9aae973..21f2a890b 100644
--- a/src/gd_topal.c
+++ b/src/gd_topal.c
@@ -1516,9 +1516,13 @@ static void free_truecolor_image_data(gdImagePtr oim)
 	int i;
 	oim->trueColor = 0;
 	/* Junk the truecolor pixels */
+#if ENABLE_FLAT_MEMORY
+	gdFree (oim->tpixels[0]);
+#else
 	for (i = 0; i < oim->sy; i++) {
 		gdFree (oim->tpixels[i]);
 	}
+#endif
 	gdFree (oim->tpixels);
 	oim->tpixels = 0;
 }
@@ -1593,11 +1597,21 @@ static int gdImageTrueColorToPaletteBody (gdImagePtr oim, int dither, int colors
 			/* No can do */
 			goto outOfMemory;
 		}
+#if ENABLE_FLAT_MEMORY
+		nim->pixels[0] = (unsigned char *) gdCalloc (sizeof (unsigned char), oim->sx * nim->sy);
+		if (!nim->pixels[0]) {
+			goto outOfMemory;
+		}
+#endif
 		for (i = 0; (i < nim->sy); i++) {
+#if ENABLE_FLAT_MEMORY
+			nim->pixels[i] = nim->pixels[0] + (i*oim->sx);
+#else
 			nim->pixels[i] = (unsigned char *) gdCalloc (sizeof (unsigned char), oim->sx);
 			if (!nim->pixels[i]) {
 				goto outOfMemory;
 			}
+#endif
 		}
 	}
 
diff --git a/windows/msys/Makefile b/windows/msys/Makefile
index d183ac25e..597e52654 100644
--- a/windows/msys/Makefile
+++ b/windows/msys/Makefile
@@ -9,7 +9,7 @@ LIBVER=3.0.1
 CDEFS=-DHAVE_ICONV -DHAVE_ICONV_H		\
 -DHAVE_ICONV_T_DEF -DHAVE_INTTYPES_H	\
 -DHAVE_STDINT_H -DHAVE_VISIBILITY -DICONV_CONST \
--DENABLE_GD_FORMATS=1
+-DENABLE_GD_FORMATS=1 -DENABLE_FLAT_MEMORY=1
 
 # GnuWin32 libs should all be installed in one place.
 GNUDIR=/c/tools/GnuWin32
